daemon/daemon.h

```diff
@@ -5,13 +5,55 @@
 
 #pragma once
 
-struct tr_error;
+#include <string>
+#include <string_view>
 
-typedef struct dtr_callbacks
+#include <libtransmission/transmission.h>
+#include <libtransmission/variant.h>
+#include <libtransmission/error.h>
+#include <libtransmission/utils.h>
+#include <libtransmission/file.h>
+#include <libtransmission/log.h>
+
+using namespace std::literals;
+
+class tr_daemon
 {
-    int (*on_start)(void* arg, bool (*setupsigfn)(void*), bool foreground);
-    void (*on_stop)(void* arg);
-    void (*on_reconfigure)(void* arg);
-} dtr_callbacks;
+#ifdef HAVE_SYS_SIGNALFD_H
+    int sigfd_ = -1;
+#endif /* signalfd API */
+    bool paused_ = false;
+    bool seen_hup_ = false;
+    std::string config_dir_;
+    tr_variant settings_ = {};
+    tr_session* my_session_ = nullptr;
+    char const* log_file_name_ = nullptr;
+    struct event_base* ev_base_ = nullptr;
+    tr_sys_file_t logfile_ = TR_BAD_SYS_FILE;
+    tr_quark key_pidfile_ = tr_quark_new("pidfile"sv);
+    tr_quark key_watch_dir_force_generic_ = tr_quark_new("watch-dir-force-generic"sv);
+
+    bool parse_args(int argc, char const** argv, bool* dump_settings, bool* foreground, int* exit_code);
+    bool reopen_log_file(char const* filename);
+    bool setup_signals();
+    void report_status();
+
+public:
+    tr_daemon() = default;
+
+    ~tr_daemon()
+    {
+#ifdef HAVE_SYS_SIGNALFD_H
+        close(sigfd_);
```

Comment:
I think this can be called when `sigfd_` is -1. Probably not a fatal error, but we ought to check before closing to avoid `EBADF`